subsys: spm: Makes reboot from non-secure configurable

sigvartmh opened the pull request:
A problem with forcing `tz_nonsecure_system_reset_req_block` to be
enabled is that it will block the debugger from reseting the bord with
soft-reset `nrfjprog -r` where pin-reset is not enabled on a board. This
relates to custom boards. By making it configurable it can be turned off
for custom boards.

Signed-off-by: Sigvart Hovland <sigvart.hovland@nordicsemi.no>

sigvartmh commented:
@joakimtoe After discussion with some of the others the default will be that it is blocked by default but now is configurable. I know you wanted it to be allowed by default as this would make it easier for customers with custom boards.

The reasoning is that it's better to enforce best practices with trust-zone having it as secure as possible. Instead of having it less secure but easier to get started with.

If it's possible to get `nrfjprog` util to do a function call into https://github.com/NordicPlayground/fw-nrfconnect-nrf/blob/f8b19084105455b10ae48a56ebeeef9748024151/subsys/spm/secure_services.c#L51-L55
this change can be omitted.

hakonfam commented on subsys/spm/spm.c:
```diff
@@ -355,7 +360,7 @@ static void spm_configure_ns(const tz_nonsecure_setup_conf_t
 	/* Set non-banked exceptions to target Non-Secure */
 	tz_nbanked_exception_target_state_set(0);
 	/* Don't allow Non-Secure firmware to issue System resets. */
```
Update comment to reflect the configurability

sigvartmh replied:
Think this is addressed now

hakonfam commented on subsys/spm/spm.c:
```diff
@@ -355,7 +360,7 @@ static void spm_configure_ns(const tz_nonsecure_setup_conf_t
 	tz_nbanked_exception_target_state_set(0);
 	/* Don't allow Non-Secure firmware to issue System resets. */
-	tz_nonsecure_system_reset_req_block(1);
+	tz_nonsecure_system_reset_req_block(SPM_BLOCK_REBOOT);
```
Could we use `IS_ENABLED(CONFIG_SPM_SERVICE_REBOOT)` directly here?

sigvartmh replied:
Would assume so. Changed.

joakimtoe commented:
@sigvartmh : Ok, since default is blocking system reset from ns, please also include a overlay to the project config that will allow it when compiling for nrf9160_pca20035?

hakonfam commented:
AFAICS what is proposed in this PR is done, if some application requires setting some other default, or if the default should change could be done in a separate PR.

ioannisg commented:
> 
> 
> @sigvartmh : Ok, since default is blocking system reset from ns, please also include a overlay to the project config that will allow it when compiling for nrf9160_pca20035?

@sigvartmh, so, production code for nrf9160_pca20035 will allow reboots from NS, _not_ via the secure service?

sigvartmh commented:
Going to close this PR, I think changes should be moved to tooling and not code.

joakimtoe commented:
@sigvartmh I dont belive this PR should have been closed? System reset behaviour should be configurable without the need to modify code? I propose reopening this since debugreset is not an option to use in all the tools..

ioannisg commented on subsys/spm/spm.c:
```diff
@@ -354,8 +353,10 @@ static void spm_configure_ns(const tz_nonsecure_setup_conf_t
+	/* Configure if Non-Secure firmware should be allowed to issue System resets or
+	 * if it should be done through a secure service.
+	 */
+	tz_nonsecure_system_reset_req_block(IS_ENABLED(CONFIG_SPM_SERVICE_REBOOT));
```
Is SPM_SERVICE_REBOOT default?

hakonfam replied:
Yes

sigvartmh replied:
Yes( separated into two Kconfigs now)

sigvartmh commented:
@joakimtoe I agree, was on vacation but tried to following the discussion thanks for reopening while I was gone :)

sigvartmh commented:
Following the discussion on Jira I think the best solution would be to have this disabled for `DEBUG` and `NO_OPTIMIZATION` targets. 

This will allow the projects to be debuggable on our Cortex-m devices with trustzone with a secure debugger. While enforcing best practice when building release candidates.

hakonfam commented:
@ioannisg , could you re-visit this?

ioannisg commented:
> 
> 
> @ioannisg , could you re-visit this?

Yeap, I will do that as soon as possible

bjda commented:
I am not entirely sure using the existing CONFIG_SPM_SERVICE_REBOOT is the best idea. If this PR is to be used as a solution to the problem of not being able to program the Thingy91, the develpoer would have to choose between
A) being able to program with current tools and 
B) Using the spm_request_system_reboot in their application

hakonfam commented:
I see your point, would you suggest having a separate independent option for allowing non-secure reset?

sigvartmh commented on subsys/spm/spm.c:
```diff
@@ -354,8 +353,10 @@ static void spm_configure_ns(const tz_nonsecure_setup_conf_t
+	/* Configure if Non-Secure firmware should be allowed to issue System resets or
+	 * if it should be done through a secure service.
+	 */
+	tz_nonsecure_system_reset_req_block(IS_ENABLED(CONFIG_SPM_SERVICE_REBOOT));
```
```suggestion
	tz_nonsecure_system_reset_req_block(CONFIG_SPM_ALLOW_NON_SECURE_RESET); 
```
Something like this @bjda  is more what you want as a purposed solution?

bjda replied:
Yes. That way, the workaround to the reset issue and configuration of the reboot service can occur independently.

bjda commented:
> 
> 
> I see your point, would you suggest having a separate independent option for allowing non-secure reset?

Yes. That way, the workaround to the reset issue and configuration of the reboot service can occur independently.

NordicBuilder commented:
Found the following issues, please fix and resubmit:

## Kconfig issues

```
Undefined Kconfig symbols:

warning: undefined symbol BOARD_NRF9160_PCA20035NS:

- Referenced at /jenkins_workspace/workspace/fw-nrfconnect-nrf_PR-1028/nrf/subsys/spm/Kconfig:64:

config SPM_ALLOW_NON_SECURE_RESET
	bool "Non-Secure reboot"
	default y if BOARD_NRF9160_PCA20035NS
	default n
	depends on IS_SPM
	help
	  This will allow applications running in Non-Secure to do a
	  system reset of the chip without going through a secure service.
	  This should be enabled for boards which doesn't have a pin-reset
	  when doing development/debugging as a debugger working in secure
	  context can't reset a device through software reset if it is blocked
	  from non-secure. Meaning you will not be able to upload new firmware
	  to the device.


```

sigvartmh commented:
> Found the following issues, please fix and resubmit:
> ## Kconfig issues
> 
> ```
> Undefined Kconfig symbols:
> 
> warning: undefined symbol BOARD_NRF9160_PCA20035NS:
> 
> - Referenced at /jenkins_workspace/workspace/fw-nrfconnect-nrf_PR-1028/nrf/subsys/spm/Kconfig:64:
> 
> config SPM_ALLOW_NON_SECURE_RESET
> 	bool "Non-Secure reboot"
> 	default y if BOARD_NRF9160_PCA20035NS
> 	default n
> 	depends on IS_SPM
> 	help
> 	  This will allow applications running in Non-Secure to do a
> 	  system reset of the chip without going through a secure service.
> 	  This should be enabled for boards which doesn't have a pin-reset
> 	  when doing development/debugging as a debugger working in secure
> 	  context can't reset a device through software reset if it is blocked
> 	  from non-secure. Meaning you will not be able to upload new firmware
> 	  to the device.
> ```

No idea why CI is complaining about the config AFIS it is defined in `boards/arm/nrf9160_pca20035/Kconfig.board`

joakimtoe commented:
> > Found the following issues, please fix and resubmit:
> > ## Kconfig issues
> > ```
> > Undefined Kconfig symbols:
> > 
> > warning: undefined symbol BOARD_NRF9160_PCA20035NS:
> > 
> > - Referenced at /jenkins_workspace/workspace/fw-nrfconnect-nrf_PR-1028/nrf/subsys/spm/Kconfig:64:
> > 
> > config SPM_ALLOW_NON_SECURE_RESET
> > 	bool "Non-Secure reboot"
> > 	default y if BOARD_NRF9160_PCA20035NS
> > 	default n
> > 	depends on IS_SPM
> > 	help
> > 	  This will allow applications running in Non-Secure to do a
> > 	  system reset of the chip without going through a secure service.
> > 	  This should be enabled for boards which doesn't have a pin-reset
> > 	  when doing development/debugging as a debugger working in secure
> > 	  context can't reset a device through software reset if it is blocked
> > 	  from non-secure. Meaning you will not be able to upload new firmware
> > 	  to the device.
> > ```
> 
> No idea why CI is complaining about the config AFIS it is defined in `boards/arm/nrf9160_pca20035/Kconfig.board`

Is it related to https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/1059/ ?

sigvartmh commented:
> > > Found the following issues, please fix and resubmit:
> > > ## Kconfig issues
> > > ```
> > > Undefined Kconfig symbols:
> > > 
> > > warning: undefined symbol BOARD_NRF9160_PCA20035NS:
> > > 
> > > - Referenced at /jenkins_workspace/workspace/fw-nrfconnect-nrf_PR-1028/nrf/subsys/spm/Kconfig:64:
> > > 
> > > config SPM_ALLOW_NON_SECURE_RESET
> > > 	bool "Non-Secure reboot"
> > > 	default y if BOARD_NRF9160_PCA20035NS
> > > 	default n
> > > 	depends on IS_SPM
> > > 	help
> > > 	  This will allow applications running in Non-Secure to do a
> > > 	  system reset of the chip without going through a secure service.
> > > 	  This should be enabled for boards which doesn't have a pin-reset
> > > 	  when doing development/debugging as a debugger working in secure
> > > 	  context can't reset a device through software reset if it is blocked
> > > 	  from non-secure. Meaning you will not be able to upload new firmware
> > > 	  to the device.
> > > ```
> > 
> > 
> > No idea why CI is complaining about the config AFIS it is defined in `boards/arm/nrf9160_pca20035/Kconfig.board`
> 
> Is it related to #1059 ?

I would assume so. Should I rebase the PR ontop of #1059?

bjda commented on subsys/spm/Kconfig:
```diff
@@ -61,6 +61,20 @@ config SPM_SERVICE_RNG
+# FIXME: Work-around for Thingy:91 until tooling issues have been worked out.
+config SPM_ALLOW_NON_SECURE_RESET
+	bool "Allow Non-Secure reset"
+	default y if BOARD_NRF9160_PCA20035NS
```
```suggestion
	default y if BOARD_NRF9160_PCA20035
```
Correct me if I'm wrong, but since this is configuration for the SPM, it is the secure board that is relevant.

sigvartmh replied:
Your are again correct. Thanks I forgot that the board define changes from each image.

sigvartmh replied:
Fixed now

sigvartmh commented:
cc: @Rallare

bjda commented on subsys/spm/spm.c:
```diff
@@ -386,8 +385,12 @@ static void spm_configure_ns(const tz_nonsecure_setup_conf_t
+	 * if it should be done through a secure service.
+	 */
+	tz_nonsecure_system_reset_req_block(
+		IS_ENABLED(CONFIG_SPM_ALLOW_NON_SECURE_RESET)
```
Is this inverted? When CONFIG_SPM_ALLOW_NON_SECURE_RESET is 1, that should NOT block reset requests.

sigvartmh replied:
That's correct(oversight from me will fix it)

sigvartmh commented on subsys/spm/Kconfig:
```diff
@@ -61,6 +61,20 @@ config SPM_SERVICE_RNG
+config SPM_ALLOW_NON_SECURE_RESET
+	bool "Allow Non-Secure reset"
+	default y if BOARD_NRF9160_PCA20035
+	default n
```
@ioannisg tracking the issue related to this PR on Jira there is a push for this being defaulted to `y` feel free to comment in Jira.

lemrey commented on subsys/spm/Kconfig:
```diff
@@ -61,6 +61,20 @@ config SPM_SERVICE_RNG
+	  This will allow the application running in Non-Secure to do a
+	  system reset of the chip without going through a secure service.
+	  This should be enabled for boards which doesn't have a pin-reset
+	  when doing development/debugging as a debugger working in secure
```
Isn't it the other way around, e.g. the debugger running in non-secure can't reset if it's blocked from secure?

sigvartmh replied:
No see Jira NRF91-313

lemrey commented on subsys/spm/Kconfig:
```diff
@@ -61,6 +61,20 @@ config SPM_SERVICE_RNG
+	help
+	  This will allow the application running in Non-Secure to do a
+	  system reset of the chip without going through a secure service.
+	  This should be enabled for boards which doesn't have a pin-reset
```
boards which s/doesn't/don't have.

sigvartmh replied:
Thanks :) Fixed

sigvartmh commented:
Split this into two commits, where one is marked with `[nrf temphack]` as it is hopefully removed when this issue is resolved but as for now we should have `allow non-secure reset` to `y` to ensure that the developer experience will be as good as possible.

lemrey commented:
I am not too fond `temphack` tags.. we don't know what is temporary, really. AFAIK there hasn't been a single `temphack` commit that has been reverted so far. I would say, simply put a nice comment, possible on a separate line for readability, 80 chars etc, about what can be changed in the future.

sigvartmh commented:
@lemrey: Squashed the commits and made a #FIXME note next to it describing what should be changed if the issue is resolved.

joakimtoe commented:
@ioannisg @carlescufi : ping

ioannisg commented on subsys/spm/Kconfig:
```diff
@@ -61,6 +61,19 @@ config SPM_SERVICE_RNG
 
+config SPM_ALLOW_NON_SECURE_RESET
+	bool "Allow Non-Secure reset"
+#FIXME: Change this to Default to `n` when the issue regarding doing a software
```
This should not be default yes, for all nRF91 boards right? We should be able to leave it to "n" for the boards not affected.

hakonfam replied:
What do you mean by "not affected"?

ioannisg replied:
>Since there are boards which do not have a pin-reset such as
	  Thingy91 this option should be enabled by default. If not a debugger
	  would not be able to issue a reset from a non-secure region even if it
	  is configure to be secure.

So, why have this "default y" for all boards, including those that have a pin reset working?

hakonfam replied:
I get your point. @sigvartmh will probably correct me, but I think we need to update this comment since this issue will hinder programming through the normal developer tools for all nrf91 boards. In which case it should be `default y` for all nrf91 boards.

Note that I might have misunderstood something, so lets wait for @sigvartmh 's response. Anyway, either the docs or the default must change.

bjda replied:
AFAIK, the reboot issue itself affects every board (see Jira NRF91-313). The JLink debugger automatically falls back to a pin reset if the soft reset doesn't work, so programming works fine in practice if pin reset is routed.

We cannot make the assumption that any new/custom board has pin reset routed, so a `y` setting until NRF91-313 is fixed avoids possibly breaking programming by default on new boards.

Adding `default n` for the nRF91 DK, which does have pin reset, would make sense.

sigvartmh replied:
@ioannisg: If you can figure a good scheme to allow it for all boards nRF91 boards without a pin-reset enabled sure, I could not. It's possible to do  `default y if nrf91xx && !nrf9160_pca10090` reasoning is that customers will create their own custom boards. We don't know if they have or have connected pin-reset as this information does not exists yet or anywhere. There is also no information in device-tree or Kconfig which gives this information-

So from supports perspective as long as the issue NRF91-313 exists it would be a lot of extra support cases with people not being able to program/reset their custom boards I assume. Maybe @Rallare could give some more input on this.

sigvartmh replied:
Added `default n if BOARD_NRF9160_PCA10090` but other defaults would have to be added on a needed basis.

sigvartmh replied:
reverted this from a security standpoint to be of for every board except Thingy as a fix will come out soon for nrfjprog.

ioannisg commented on subsys/spm/Kconfig:
```diff
@@ -71,6 +71,20 @@ config SPM_SERVICE_READ
+	  This will allow the application running in Non-Secure to do a
+	  system reset of the chip without going through a secure service.
+	  Since there are supported boards which do not have a pin-reset such
+	  as Thingy91 this option should be enabled by default these boards.
```
on* these boards

ioannisg commented on subsys/spm/Kconfig:
```diff
@@ -71,6 +71,20 @@ config SPM_SERVICE_READ
+	  system reset of the chip without going through a secure service.
+	  Since there are supported boards which do not have a pin-reset such
+	  as Thingy91 this option should be enabled by default these boards.
+	  If not enabled a  debugger would not be able to issue a reset from a
```
double space

ioannisg commented on subsys/spm/Kconfig:
```diff
@@ -71,6 +71,20 @@ config SPM_SERVICE_READ
+	  Since there are supported boards which do not have a pin-reset such
+	  as Thingy91 this option should be enabled by default these boards.
+	  If not enabled a  debugger would not be able to issue a reset from a
+	  non-secure region even if it is configure to be secure.
```
what is non-secure "region" here?

bjda replied:
A more precise phrasing might be 
```suggestion
	  If not enabled, a debugger will not be able to issue a soft reset
	  while the core is executing non-secure code.
```

sigvartmh replied:
Added phrase about non-secure code.

joakimtoe commented:
@ioannisg What is blocking this PR?

Rallare commented on subsys/spm/Kconfig:
```diff
@@ -71,6 +71,20 @@ config SPM_SERVICE_READ
+	bool "Allow Non-Secure reset"
+#FIXME: Change this to Default to `n` when the issue regarding doing a software
+#	reset from non-secure state with a secure debugger is resolved.
+	default y if BOARD_NRF9160_PCA20035
```
I would suggest that both secure and non-secure are able to trigger soft-reset. Any users that require a soft-reset to go through secure region shall know what they are doing, and should then add this config entry manually.

My suggestion is to set this default y in all cases, allowing both regions to execute soft-reset.

sigvartmh replied:
@ioannisg any objections?

ioannisg replied:
This sort of contradicts the long-term strategy of least-privilege policy. For example, we allocate to non-secure only the resources we are using, right? Same fore reset; this is, really, a secure side thing: Non-Secure should not be able to reboot the SoC, if you we don't know what the Secure side is doing. So it's OK to have this as configurable option (not ideal), but not default-y.

bjorn-tore-taraldsen replied:
Right now we need to look at usability vs security. It doesn't help how secure we makes our examples (by default configuration) if it leads to our customer having problem running/ debugging it. We should have a solution that can be used by our customer without having to use DevZone /Support.
My take on this is make user friendly bu default and add documentation describing what you should have when going into production.

ioannisg commented on subsys/spm/Kconfig:
```diff
@@ -71,6 +71,20 @@ config SPM_SERVICE_READ
 
+config SPM_ALLOW_NON_SECURE_RESET
+	bool "Allow Non-Secure reset"
+#FIXME: Change this to Default to `n` when the issue regarding doing a software
```
You should remove the FixMe comment (perhaps add a minor inline Fixme in the default y sentence, pointing to a Jira/Github issue

sigvartmh replied:
Moved.

sigvartmh commented:
## Change
* Set default to `y` for allowing non secure resets.
* Rephrased help text for Kconfig
* Added paragraph about allowing non-secure reset
* Added fixme tag for both doc and Kconfig text.

sigvartmh commented:
Currently broken a bit I'll fix it up ASAP

sigvartmh commented on include/secure_services.rst:
```diff
@@ -17,4 +20,4 @@ API documentation
 
 .. doxygengroup:: secure_services
    :project: nrf
-   :members:
```
Added unix fileending to file.

ioannisg commented on subsys/spm/Kconfig:
```diff
@@ -71,6 +71,16 @@ config SPM_SERVICE_READ
 
+config SPM_ALLOW_NON_SECURE_RESET
+	bool "Allow Non-Secure reset"
+	default y
```
Assuming that we agree to do like this (i.e. allow by default a Non-Secure call to NVIC_SystemReset()), which I really think we do not need to, since there are lots' of options to address this, e.g. workaround only for Thingy, or only for boards without PIN Reset capability, etc....

...we may need to fix the alignment with SPM_SERVICE_REBOOT. These two, can't be default y at the same time (unless I am missing something here). That is, if we don't lock the Non-Secure system reset, what is the point of setting the secure service to reboot?

sigvartmh replied:
SPM_SERVICE_REBOOT does not interfere with the ability of doing NVIC_SystemReset() so they can both be enabled at the same time.

The point of having secure service reboot with this enabled would be if you have written code which uses the secure service but you want to use a debugger to debug your code. This means you want to have the API exposed so that your code can use it but  you also want to leave it open so that the debugger can reset the device. At least that's a use-case I can see.

ioannisg replied:
SPM_ALLOW_NON_SECURE_RESET is not a debugger-specific option. Neither the name, nor the actual functionality suggests it.

Setting SPM_ALLOW_NON_SECURE_RESET allows the NS firmware to directly call NVIC_SystemReset() in the code. So we can't be having a parallel Secure Service that instructs the Secure firmware to call NVIC_SystemReset(). We need to sort this out because it will be confusing, and, in fact, not needed. It will consume flash, ram and execution resources, for no reason.

Have you looked at the definition of SPM_SERVICE_REBOOT?

```
config SPM_SERVICE_REBOOT
	bool "Request a reboot"
	default y
	select REBOOT
	help
	  The Non-Secure Firmware is not allowed to reboot the chip. This
	  service allows it to request a reboot from the SPM.
```

So, it claims that if the option is enabled, the non-secure FW can't issue a system reset, which is clearly, wrong.

Kconfig options are there, so we and product developers may use them to configure our software. We can't be having contradicting options enabled.

sigvartmh replied:
@ioannisg :
> SPM_ALLOW_NON_SECURE_RESET is not a debugger-specific option. Neither the name, nor the actual functionality suggests it.

Because it's not, I was merely stating a use-case not the pure rationale behind the option which has been discussed in the JIRA issue and this PR. It has to do with devices which do not support pin-reset etc. Also, the options include the following help text, which is also the functionality which it coincidentally toggles on and off.

```
If not enabled, a debugger will not be able to issue a soft reset while the
core is executing Non-Secure code.
```

> Have you looked at the definition of SPM_SERVICE_REBOOT?

Since this PR has been bouncing back and forth on the issue I abstained from updating the help text for other options, and I would be careful to do so before we can agree on what should be implemented.

> So, it claims that if the option is enabled, the non-secure FW can't issue a system reset, which is clearly, wrong.

The statement is wrong as you said and the configuration would be ambiguous if both allow non-secure reset and `SPM_SERVICE_REBOOT` are enabled which of course is never a good thing, nice spot.

I do, however, wonder what the best scheme of having this configurable would be any suggestions? One option is to change the text of the option you listed to:

```
If the Non-Secure Firmware is not allowed to reboot the chip.
This service will allow it to request a reboot from the SPM.
```

This way, the configurations would not be contradicting. Of course, this does not address the issue by having both enabled you'll consume a few more bytes of flash. By default, I do agree both should not be enabled, but I don't think they should be coupled where the rationale has been mentioned before here:

> I am not entirely sure using the existing CONFIG_SPM_SERVICE_REBOOT is the best idea. If this PR is to be used as a solution to the problem of not being able to program the Thingy91, the develpoer would have to choose between
> A) being able to program with current tools and
> B) Using the spm_request_system_reboot in their application

Our current predicament is that we can't have secure and ease of use chose both. In this way, I like @pizi-nordic suggestion of having this "HARDENED MODE" where product developers who are focused on security can turn it on. However we should be careful of claiming too much also that would be out-of-scope for this PR.

I also think both support @Rallare and maintainers @ioannisg should look revisit this commit and see if it can be reverted, when a workaround for the issue with regards to reset has been resolved in tooling.

As for now, I see the argument of having something working with an opened attack vector is stronger than something that does not work for product developer as long as it is communicated how and why they should close the attack vector when releasing their product.

sigvartmh commented:
Change log:
* Renamed allow to **block**
* Removed not operator from `IS_ENABLED`
* Reworked help text for both Secure service reboot and block system reset non-secure.
* Moved things around in menuconfig.

Here is a current look at the menuconfig where i moved secure services out of its own config three(menu selector) so that the options are more visible for the users in case they want to block system reset from non-secure.
![spm_menuconfig](https://user-images.githubusercontent.com/3646850/65616901-19ba5b80-dfbc-11e9-9a30-a732fa13d819.PNG)

More renaming of the secure services could be required. This is only a proposal so if anyone has a better idea for naming, text and structure please feel free to share.

sigvartmh commented on subsys/spm/Kconfig:
```diff
@@ -69,11 +65,24 @@ config SPM_SERVICE_READ
-	bool "Boot silently"
-	default n
+config SPM_BLOCK_NON_SECURE_RESET
+	bool "Block system reset from Non-Secure"
```
```suggestion
	bool "Block system reset from Non-Secure domain"

sigvartmh replied:
fixed

sigvartmh commented:
> Please, see if the option could be turned on only for the platforms it needs to be

Could you revisit this? Do you find the configurations more coherent now together with their help text?

carlescufi commented:
@sigvartmh could you please get CI passing with @thst-nordic's help? Then I can merge it.

sigvartmh commented:
Sure I'll try I'll just have a look into the doc error 1st

hakonfam commented:
Ping @ioannisg  looks like you have a stale review

sigvartmh commented:
@jorgenmk and @lemrey Could you look over the changes to lwm2m_carrier and see if it makes sens?
